lib/memoize.js

```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+const generateKey = args => args
+  .map(item =>
+    `${item.toString()}:${typeof item}`)
+  .join('|');
+
+const promisifyClosure = (fn, hasher) => (...args) => new Promise(
+  (resolve, reject) => {
+    fn(...args, (err, data) => {
+      const res = hasher(err, data) || data;
+      return (err ? reject(err) : resolve(res));
+    });
+  });
+
+const asyncMemoize = (fn, config = {}) => {
+  if (typeof fn !== 'function') {
+    throw new Error('fn is not a function');
+  }
+
+  const { isCb, cacheSize } = config;
+  const cache = new Map();
+
+  const memoized = (...args) => {
+    let wrappedFn = fn;
+    const key = generateKey(args);
+    if (isCb) {
+      const lastArg = args[args.length - 1];
```

Comment:
```suggestion
      const lastArg = args.unshift();
```

Reply:
I think it's better not to do that, because we immediately modify the argument array, and if we were not passed a callback (although the function follows this contract), we would lose one argument for nothing.

Reply:
I was mistaken, you are right